### ISSUE TITLE
cmake generate step failed on linux

### ISSUE THREAD

**Issue by ehuaa** [iree-org/iree#12640]
### What happened?

**When I followed the build instructions on the website using command as follows** 

cmake -GNinja -B ../iree-build/ -S .     
-DCMAKE_BUILD_TYPE=RelWithDebInfo     
-DIREE_ENABLE_ASSERTIONS=ON     
-DCMAKE_C_COMPILER=clang     
-DCMAKE_CXX_COMPILER=clang++     
-DIREE_ENABLE_LLD=ON     
-DCMAKE_C_COMPILER_LAUNCHER=ccache     
-DCMAKE_CXX_COMPILER_LAUNCHER=ccache

**it turns out a failure of missing variable and cmake generate step failed.**

-- Looking for a CUDA compiler
-- Looking for a CUDA compiler - /usr/local/cuda/bin/nvcc
-- The CUDA compiler identification is NVIDIA 11.4.152
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- IREE custom_dispatch/vulkan/shaders ignored -- glslc not found
-- Configuring done
**CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_CUDA_COMPILE_SEPARABLE_COMPILATION
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_CUDA_COMPILE_SEPARABLE_COMPILATION
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.**

**my nvcc version is**

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Mon_Oct_11_21:27:02_PDT_2021
Cuda compilation tools, release 11.4, V11.4.152
Build cuda_11.4.r11.4/compiler.30521435_0

**Can you help me fix this correct, thanks~**

### Steps to reproduce your issue
git clone to my own pc and 
cmake -GNinja -B ../iree-build/ -S .     
-DCMAKE_BUILD_TYPE=RelWithDebInfo     
-DIREE_ENABLE_ASSERTIONS=ON     
-DCMAKE_C_COMPILER=clang     
-DCMAKE_CXX_COMPILER=clang++     
-DIREE_ENABLE_LLD=ON     
-DCMAKE_C_COMPILER_LAUNCHER=ccache     
-DCMAKE_CXX_COMPILER_LAUNCHER=ccache

### What component(s) does this issue relate to?

Other

### Version information

the newest version cloned

**Comment by ehuaa** [comment]
I solved this problem by add set(CMAKE_CUDA_COMPILE_SEPARABLE_COMPILATION ON)
to CMakeLists.txt.

**Comment by ScottTodd** [comment]
You might also want to update Ninja and/or CMake.

```
CMake$ git log -S "CMAKE_CUDA_COMPILE_SEPARABLE_COMPILATION"
commit f7c6b9833c55257d2b83e2cb5678139ba52df3dc
Author: Robert Maynard <rmaynard@nvidia.com>
Date:   Wed Mar 30 15:12:25 2022 -0400

    CUDA: Ninja generator generates valid compile database
    
    The ninja generator has been updated to poperly write the compile
    database when the CUDA language has been enabled.
    
    Fixes #23368
```

That points to https://gitlab.kitware.com/cmake/cmake/-/issues/23368, which looks similar to what you saw.
